### PR TITLE
Fix loading binary resources with float=64

### LINKED ISSUE #59320: Unable to load float=32 created binary resource  on float=64
### Godot version

db5271dd34ea2b1894e1e2cf9ee9519c84142608 (closest)

### System information

Fedora 35, Nvidia 3000 series

### Issue description

Godot cannot load scn in float=64 when they're from float=32. Errors for Vector\<Vector2>, Vector\<Vector3>, Vector\<Color>.

`Vector2 size is NOT 8!"`

```c++
	case VARIANT_PACKED_VECTOR2_ARRAY: {
			uint32_t len = f->get_32();

			Vector<Vector2> array;
			array.resize(len);
			Vector2 *w = array.ptrw();
			if (sizeof(Vector2) == 8) {
				f->get_buffer((uint8_t *)w, len * sizeof(real_t) * 2);
#ifdef BIG_ENDIAN_ENABLED
				{
					uint32_t *ptr = (uint32_t *)w.ptr();
					for (int i = 0; i < len * 2; i++) {
						ptr[i] = BSWAP32(ptr[i]);
					}
				}

#endif

			} else {
				ERR_FAIL_V_MSG(ERR_UNAVAILABLE, "Vector2 size is NOT 8!");
			}

			r_v = array;

		} break;
```

### Steps to reproduce

Builds of godot with float=64 fail to load binary serialized PackedScene defined as float=32.

1. create a resource with these types in godot float=32
2. open in float=64
3. See variant error

### Minimal reproduction project

[Double Import Game Project.zip](https://github.com/godotengine/godot/files/8310434/Double.Import.Game.Project.zip)

### PR DESCRIPTION
**Pull request by Zylann**
Fixes #59320

- PackedVector2Array
- PackedVector3Array
- PackedColorArray

### CODE REVIEW

File: core/io/resource_format_binary.cpp [hunk 1]
```diff
@@ -579,7 +600,9 @@ Error ResourceLoaderBinary::parse_variant(Variant &r_v) {
 			array.resize(len);
 			Color *w = array.ptrw();
 			if (sizeof(Color) == 16) {
-				f->get_buffer((uint8_t *)w, len * sizeof(real_t) * 4);
+				// Colors always use `float` even with double-precision support enabled
+				static_assert(sizeof(Color) == 4 * sizeof(float));
```
Comment (akien-mga) [inline core/io/resource_format_binary.cpp:604]: There's `if (sizeof(Color) == 16)` before so this would never be true.

**Comment by fire** [comment]
The forward version of old float=32 to float=64 works. Need test the reverse.

File: core/io/resource_format_binary.cpp (hunk 1 above)
Comment (Zylann) [reply core/io/resource_format_binary.cpp:604]: This is a `static_assert`, it comes up at compile-time. Maybe we can remove the `if` then?

**Comment by Zylann** [comment]
@fire how do we know it's related, if the scene didn't even load before?

**Comment by fire** [comment]
Sorry for the delay. Looking into this.

**Review by akien-mga — APPROVED** [review]
Confirmed OK in PR review meeting by @reduz and @lawnjelly.

File: core/io/resource_format_binary.cpp [hunk 2]
```diff
@@ -101,6 +101,51 @@ void ResourceLoaderBinary::_advance_padding(uint32_t p_len) {
...
+				}
+			}
+#endif
+		} else if (sizeof(real_t) == 4) {
+			// May be slower, but this is for compatibility. Eventually the data should be converted.
+			for (size_t i = 0; i < count; ++i) {
+				dst[i] = f.get_double();
+			}
+		} else {
+			// Wtf
```
[the remark this answers is not in the document]
Comment (Zylann) [reply core/io/resource_format_binary.cpp]: Removed the comment

**Review by aaronfranke — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**